src/relay/backend/contrib/dnnl/query_layout.cc

```diff
@@ -0,0 +1,340 @@
+  const format_kind_t blocked = dnnl_blocked;
+  if (!(md->data.format_kind == blocked)) {
+    array_set(blocks, 0, md->data.ndims);
+    return;
+  }
+  array_set(blocks, 1, md->data.ndims);
+  const auto& bd = md->data.format_desc.blocking;
+  for (int iblk = 0; iblk < bd.inner_nblks; ++iblk)
+    blocks[bd.inner_idxs[iblk]] *= bd.inner_blks[iblk];
+}
+
+inline bool has_runtime_strides(const dnnl::memory::desc* md) {
+  using format_kind_t = dnnl_format_kind_t;
+  const format_kind_t blocked = dnnl_blocked;
+  if (!(md->data.format_kind == blocked)) return false;
+  for (int d = 0; d < md->data.ndims; ++d)
+    if (md->data.format_desc.blocking.strides[d] == DNNL_RUNTIME_DIM_VAL) return true;
+  return false;
+}
+
+std::string md2fmt_tag_str(const dnnl::memory::desc* md) {
+  const auto& blk = md->data.format_desc.blocking;
+
+  dims_t blocks = {0};
+  compute_blocks(blocks, md);
+
+  char dim_chars[DNNL_MAX_NDIMS + 1];
+
+  dims_t ou_blocks = {0};
+  array_copy(ou_blocks, md->data.padded_dims, md->data.ndims);
+
+  bool plain = true;
+  for (int d = 0; d < md->data.ndims; ++d) {
+    dim_chars[d] = (blocks[d] == 1 ? 'a' : 'A') + static_cast<char>(d);
+    if (blocks[d] != 1) plain = false;
+    ou_blocks[d] /= blocks[d];
+  }
+
+  // Can't report meaningful tag for runtime dimensions.
+  if (has_runtime_strides(md)) return "*";
+
+  dims_t strides;
+  array_copy(strides, blk.strides, md->data.ndims);
+
+  simultaneous_sort(strides, ou_blocks, dim_chars, md->data.ndims,
+                    [](dim_t a, dim_t b) { return b - a; });
+
+  dim_chars[md->data.ndims] = '\0';
+
+  std::string s(dim_chars);
+
+  if (!plain) {
+    for (int iblk = 0; iblk < blk.inner_nblks; ++iblk) {
+      char c = ('a' + static_cast<char>(blk.inner_idxs[iblk]));
+      s += (std::to_string(blk.inner_blks[iblk]) + c);
+    }
+  }
+  return s;
+}
+
+dnnl::memory::dims str2dims(std::string str_shape, int input_size) {
+  std::string str_reg = "(\\d*)";
+  for (int i = 0; i < input_size - 1; i++) {
+    str_reg.append(",(\\d*)");
+  }
+  std::regex rex(str_reg);
+  std::smatch m;
+  dnnl::memory::dims out_dims;
+  if (std::regex_search(str_shape, m, rex)) {
+    std::transform(m.begin() + 1, m.end(), std::back_inserter(out_dims),
+                   [](const std::string& str) { return std::stoi(str); });
+  } else {
+    LOG(FATAL) << "Unsupported shape for querying optimal dnnl layout: " << str_shape;
+  }
+  return out_dims;
+}
+
+std::string get_optimal_layout_for_conv(int input_size, std::string weight_shape,
+                                        std::string out_shape, std::string paddings,
+                                        std::string strides, std::string dilates, std::string G) {
+  dnnl::engine eng(dnnl::engine::kind::cpu, 0);
+  dnnl::stream s(eng);
+  using tag = dnnl::memory::format_tag;
+  using dt = dnnl::memory::data_type;
+
+  dnnl::memory::dim groups = std::stoi(G);
+  dnnl::memory::dims weight_dims_ = str2dims(weight_shape, input_size);

```

Comment:
Check if it's valid for the input args, like num of dims = 5 or 4 for weight?

Reply:
This function supports `conv1d`, `conv2d`, `conv3d` and even with `groups>1`, so 6 / 5 / 4 / 3 dims input is valid.

Reply:
As this function accepts the string args passing from python front-end, better to validate these args before further processing on them.

Reply:
1. add checks to validate the format of the input shapes as `[1-9]*(, [1-9])*`
2. add `data_layout` and `kernel_layout` into the input list to validate the input shapes.